core/internal/filetransfer/file_transfer_azure.go

```diff
@@ -0,0 +1,416 @@
+	onceVal, _ := ft.containerClients.once.LoadOrStore(containerName, &sync.Once{})
+	once := onceVal.(*sync.Once)
+	once.Do(func() {
+		cred, err := azidentity.NewDefaultAzureCredential(nil)
+		if err != nil {
+			ft.logger.Error("Unable to create Azure credential", "err", err)
+			return
+		}
+		client, err := container.NewClient(containerName, cred, nil)
+		if err != nil {
+			ft.logger.Error("Unable to create Azure client", "err", err)
+			return
+		}
+		ft.containerClients.clients.Store(containerName, client)
+	})
+}
+
+// Upload uploads a file to the server.
+func (ft *AzureFileTransfer) Upload(task *ReferenceArtifactUploadTask) error {
+	ft.logger.Debug("Azure file transfer: uploading file", "path", task.PathOrPrefix)
+
+	return nil
+}
+
+type ParsedBlobInfo struct {
+	AccountUrl string
+	Container  string
+	BlobPrefix string
+}
+
+// Download downloads a file from the server.
+func (ft *AzureFileTransfer) Download(task *ReferenceArtifactDownloadTask) error {
+	ft.logger.Debug(
+		"Azure file transfer: downloading file",
+		"path", task.PathOrPrefix,
+		"ref", task.Reference,
+	)
+
+	// Parse the reference path to get the account URL and blob path
+	accountUrl, fullBlobPath, err := parseCloudReference(task.Reference, azureScheme)
+	if err != nil {
+		return ft.formatDownloadError("error parsing reference", err)
+	}
+	pathSplit := strings.SplitN(fullBlobPath, "/", 2)
+	fullAccountUrl := fmt.Sprintf("%s://%s", azureScheme, accountUrl)
+	blobInfo := ParsedBlobInfo{
+		AccountUrl: fullAccountUrl,
+		Container:  pathSplit[0],
+		BlobPrefix: pathSplit[1],
+	}
+
+	// Setup the client if it is not already set up
+	ft.SetupClient(fullAccountUrl)
+	_, ok := ft.clients.clients.Load(fullAccountUrl)
+	if !ok {
+		return ft.formatDownloadError(
+			"error setting up Azure account client",
+			fmt.Errorf("client not found"),
+		)
+	}
+
+	var blobNames []string
+	if task.HasSingleFile() {
+		blobName, versionId, err := ft.getBlobName(blobInfo, task)
+		if err != nil {
+			return ft.formatDownloadError(
+				"error getting correct blob name and version",
+				err,
+			)
+		}
+		if versionId != "" {
+			err = task.SetVersionID(versionId)
+			if err != nil {
+				return ft.formatDownloadError(
+					"error setting version ID",
+					err,
+				)
+			}
+		}
+		blobNames = []string{blobName}
+	} else {
+		blobNames, err = ft.listBlobsWithPrefix(blobInfo)
+		if err != nil {
+			return ft.formatDownloadError(
+				fmt.Sprintf("error finding blobs with prefix %s", blobInfo.BlobPrefix),
+				err,
+			)
+		}
+	}
+
+	err = ft.downloadFiles(blobInfo, blobNames, task)
+	if err != nil {
+		return ft.formatDownloadError(
+			fmt.Sprintf("error downloading reference %s", task.Reference),
+			err,
+		)
+	}
+	return nil
+}
+
+// getBlobName tries to get the blob name and version ID that matches the
+// expected digest for the given task.
+func (ft *AzureFileTransfer) getBlobName(
+	blobInfo ParsedBlobInfo,
+	task *ReferenceArtifactDownloadTask,
+) (string, string, error) {
```

Comment:
i think the same comment i have for `getCorrectBlobVersion` is valid here as well re: code organization.